{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ca01ca05-25b9-4ca1-850e-81d942c9cce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### IMPORT LIBS AND FXS ####\n",
    "import os, sys, glob, subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import ticker\n",
    "import re\n",
    "\n",
    "# Supress warnings\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e0bfdf-b517-4ff8-b838-6a709b4c5120",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### ALIGN AND CALL PEAKS ####\n",
    "\n",
    "## PARAMS\n",
    "# DATA\n",
    "data_dir = \"/groups/doudna/projects/mtrinidad_projects/ChIPSeq_HSKW/KMW_060324_ChIP/FASTQS/\"\n",
    "r1_suffix = \"_L001_R1_001.fastq.gz\"\n",
    "r2_suffix = \"_L001_R2_001.fastq.gz\"\n",
    "r1s = [x for x in glob.glob(f\"{data_dir}*/*{r1_suffix}\", recursive = True)]\n",
    "bases = list(set([re.split(r\"_S[\\d]*_\", x.split(\"/\")[-1])[0] for x in r1s]))\n",
    "\n",
    "# ALIGNMENTS\n",
    "bam_suffix = \"_sorted.bam\"\n",
    "bam_dir = f\"/groups/doudna/projects/mtrinidad_projects/ChIPSeq_HSKW/KMW_060324_ChIP/Alignments/\"\n",
    "if not os.path.exists(bam_dir):\n",
    "    os.mkdir(bam_dir)\n",
    "\n",
    "# PEAK-CALLING\n",
    "peak_dir = \"/groups/doudna/projects/mtrinidad_projects/ChIPSeq_HSKW/KMW_060324_ChIP/Peaks/\"\n",
    "if not os.path.exists(peak_dir):\n",
    "    os.mkdir(peak_dir)\n",
    "macs2_dir = \"/home/mtrinidad/miniconda3/envs/ChIP/bin/macs2\"\n",
    "\n",
    "# THREADS\n",
    "threads = \"$SLURM_CPUS_ON_NODE\"\n",
    "\n",
    "# REFERENCES\n",
    "ref_dir = \"/groups/doudna/team_resources/shared_databases/human_reference_genomes/hg38/GENCODE_GRCh38p14_v44/GRCh38_PRI_bt\"\n",
    "\n",
    "# BACKGROUND SUBTRACTION BAMS: exploratory peak calling setup (Publication uses pseudoreplicates to explore enrichment. See PseudoRep_and_RPKM.ipynb)\n",
    "background_sample_dict = {\n",
    "                          'ChIP__KMW_060324_dSpRYapo_UCLA':[f\"{bam_dir}ChIP__KMW_060324_dSpRYapo_input_UCLA{bam_suffix}\"],\n",
    "                          'ChIP__KMW_060324_dSpRY_UCLA':[f\"{bam_dir}ChIP__KMW_060324_dSpRYapo_input_UCLA{bam_suffix}\"],\n",
    "                          'ChIP__KMW_060324_dWTapo_UCLA':[f\"{bam_dir}ChIP__KMW_060324_dWTapo_input_UCLA{bam_suffix}\"],\n",
    "                          'ChIP__KMW_060324_dWT_UCLA':[f\"{bam_dir}ChIP__KMW_060324_dWTapo_input_UCLA{bam_suffix}\"]\n",
    "                        }\n",
    "\n",
    "# SBATCH HEADER AND BASH SCRIPT\n",
    "bash_script = \"Alignments_and_Peak_Calling.sh\"\n",
    "bash_lines = [\"#!/bin/bash\", \"#SBATCH -p standard\", \"#SBATCH --job-name ALIGN_MACS2\", \n",
    "               \"#SBATCH -o %j.out\", \"#SBATCH -e %j.err\", \"## ACTIVATE CONDA\", \n",
    "              'eval \"$(conda shell.bash hook)\"', 'conda activate ChIP']\n",
    "\n",
    "\n",
    "## ALIGN WITH BOWTIE\n",
    "bams = []\n",
    "bash_lines.append(\"#### ALIGN WITH BOWTIE ####\")\n",
    "print(\"#### ALIGN WITH BOWTIE #### \")\n",
    "for r1 in r1s:\n",
    "    base = re.split(r\"_S[\\d]*_\", r1.split(\"/\")[-1])[0]\n",
    "    sample_id = re.split(r1_suffix, r1.split(\"/\")[-1])[0]\n",
    "    r2 = r1.replace(r1_suffix, r2_suffix)\n",
    "    sorted_bam = f\"{bam_dir}{sample_id}{bam_suffix}\"\n",
    "    bams.append(sorted_bam)\n",
    "    bash_lines.append(f\"## {sample_id}\")\n",
    "    bowtie_align = (f\"bowtie2 -x {ref_dir} -1 {r1} -2 {r2} -p {threads} \"\n",
    "                    f\"--no-discordant --local --no-mixed --maxins 1000 | \" # Minimize large gaps (confuses MACS2)\n",
    "                    f\"samtools view -bSh - | samtools sort -@ {threads} -o {sorted_bam}\")\n",
    "    bash_lines.append(bowtie_align)\n",
    "    index_str = f\"samtools index {sorted_bam} -@ {threads}\"\n",
    "    bash_lines.append(index_str)\n",
    "\n",
    "## CALL PEAKS\n",
    "bash_lines.append(f\"#### CALL PEAKS ####\")\n",
    "p_val = 0.01\n",
    "print(f\"#### CALL PEAKS ####\", bams)\n",
    "for bam in bams:\n",
    "    if \"input_\" in bam:\n",
    "        continue\n",
    "    base = re.split(r\"_S[\\d]*_\", bam.split(\"/\")[-1])[0]\n",
    "    sample_id = re.split(bam_suffix, bam.split(\"/\")[-1])[0]\n",
    "    input_ctrl_bam = background_sample_dict[base]\n",
    "    bash_lines.append(f\"## {sample_id}\")\n",
    "    print(f\"## {sample_id}\")\n",
    "    \n",
    "    macs2_peaks = (f\"{macs2_dir} callpeak -t {bam} -c {' '.join(input_ctrl_bam)} \"\n",
    "                   f\"-g hs -n {sample_id} -f BAM -B --SPMR -p {p_val} --outdir {peak_dir}  \"\n",
    "                   f\"2> {peak_dir}{sample_id}_macs2.log\")\n",
    "    bash_lines.append(macs2_peaks)\n",
    "\n",
    "## MAKE BIGWIGS\n",
    "bash_lines.append(f\"#### MAKE BIGWIGS ####\")\n",
    "dedup_bws = []\n",
    "\n",
    "for bam in bams:\n",
    "    base = re.split(r\"_S[\\d]*_\", bam.split(\"/\")[-1])[0]\n",
    "    sample_id = re.split(bam_suffix, bam.split(\"/\")[-1])[0]\n",
    "    bash_lines.append(f\"## {sample_id}\")\n",
    "    \n",
    "    # CPM DEDUPLICATED\n",
    "    dedup_bw = f\"{bam_dir}{sample_id}_CPM_Norm_DEDUP.bigWig\"\n",
    "    bw_dedup = (f\"bamCoverage -b {bam} -o {dedup_bw} --normalizeUsing CPM \"\n",
    "                f\"--numberOfProcessors {threads} --ignoreDuplicates --binSize 1\")\n",
    "    bash_lines.append(bw_dedup)\n",
    "    \n",
    "    # RAW WITH DUPLICATES\n",
    "    dup_raw_bw = f\"{bam_dir}{sample_id}_RAW_DUP.bigWig\"\n",
    "    bw_dup = (f\"bamCoverage -b {bam} -o {dup_raw_bw} \"\n",
    "                    f\"--numberOfProcessors {threads} --binSize 1\")\n",
    "    bash_lines.append(bw_dup)\n",
    "    \n",
    "\n",
    "## CORRELATION\n",
    "bash_lines.append(f\"#### CORRELATION ####\")\n",
    "\n",
    "# GET NPZ\n",
    "npz = f\"{bam_dir}MultiBamSummary_vDeDupBWs.npz\"\n",
    "counts = f\"{bam_dir}MultiBamSummary_Counts_vDeDupBWs.npz\"\n",
    "bw_suffix = \"_RAW_DEDUP.bigWig\"\n",
    "labels = [x.split(\"/\")[-1].replace(bw_suffix, \"\") for x in dedup_bws]\n",
    "\n",
    "make_npz = (f'multiBigwigSummary bins -b {\" \".join(dedup_bws)} --numberOfProcessors {threads} '\n",
    "            f'--labels {\" \".join(labels)} --outFileName {npz}')\n",
    "bash_lines.append(make_npz)\n",
    "\n",
    "# PLOT\n",
    "get_corr = (f\"plotCorrelation -in {npz} \"\n",
    "            \"--corMethod pearson \"\n",
    "            \"--skipZeros \"\n",
    "            \"--plotTitle 'Pearson Correlation of BAM Coverage' \"\n",
    "            \"--whatToPlot heatmap \"\n",
    "            \"-o DeDupBWs_PearsonCorr_Scores.png \"\n",
    "            \"--outFileCorMatrix DeDupBWs_PearsonCorr_Scores.tab \"\n",
    "            \"--plotNumbers \"\n",
    "            \"--colorMap RdYlBu \" \n",
    "            \"--removeOutliers \" \n",
    "        )\n",
    "bash_lines.append(get_corr)\n",
    "\n",
    "# WRITE FILE\n",
    "with open(bash_script, \"w\") as f:\n",
    "    f.writelines(\"\\n\".join(bash_lines))  \n",
    "\n",
    "\n",
    "#os.system(f\"sbatch {bash_script}\") # JID:471871"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "903f0a0f-2063-4f01-b58d-5ad0822e477f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Test_Py_3_9",
   "language": "python",
   "name": "test_py_3_9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
